### DISCUSSION TITLE
Ship Systems - Shields

### DISCUSSION THREAD

**Discussion by alexanderson1993** [Thorium-Sim/thorium-nova#281] [Feature Proposals]
_For more information about the ship systems discussions, see https://github.com/Thorium-Sim/thorium-nova/discussions/224_

---

# Shields

Shields are a kind of force field used to protect the hull of the ship from damage. Possible dangers include energy and projectile weapons and small objects the ship collides with in travel.

Shields operate by pulsing space-magic energy at certain frequencies. Different frequencies can block different things better, much like noise-cancelling headphones. This includes phasers, which are fired at certain frequencies, different kinds of radiation, and other forms of harmonic energy.

Shields are projected from the hull of the ship in a bubble shape, with 6 distinct sections, each with their own power consumption, strength, efficiency level, and frequency: fore, aft, port, starboard, dorsal, ventral. These shield sections can be controlled altogether in tandem, or individually, as needed. This includes raising, lowering, and changing the frequency.

Shields can either be raised or lowered. For shields to raise, they require a substantial amount of power. If power is limited to them, they will raise slowly; otherwise they will take as much power as is available, perhaps disrupting other systems in the process. Lowering the shields dissipates the excess power.

While raised, shields consume a relatively small amount of power to stay raised. However, whenever they deflect something, that will decrease their overall strength (measured in percent from 0 - 100%), which will cause the shields to draw power to regenerate themselves. With enough time between impacts and sufficient power, shields could theoretically last forever - with a big caveat.

Every deflection also decreases the _efficiency_ of the shields system slightly, which is different from the shield strength. The more efficiency decreases, the longer it takes for shields to regenerate given the same amount of power. Maintenance teams would have to make repairs to restore efficiency.

This gives much more significance to the cliché "Divert power to the shields!" Diverting power to shields would increase their regeneration rate, which avoids having them fail at some point.

When shields are lowered, the ship isn't entirely defenseless. The ships hull is made of space-magic material that can withstand several energy or projectile weapon impacts. However, the hull cannot be repaired or regenerated easily, while the shields can - reminiscent of the player's shields in the Halo franchise. That's why raising the shields is the #1 tactical moves that will help you win a fight.

### Practical Considerations

Shields (along with the non-regenerating hull) serve as the hit-points of the ship. When enemy weapons impact the crew's ship, the game will calculate which shields were impacted and deduct some amount of strength based on a few factors:

- The yield of the weapon
- The frequency of the weapon
- The frequency of the shields
- Other attributes about the weapon that we might add, such as blast radius, EMP ordinances, etc.

As long as shield strength remains above 0%, the hull strength will not be affected by weapons fire. Any damage sustained past the shields reaching 0% will be applied directly to the hull strength. However, shields will continue to regenerate as long as they are raised and have sufficient power. Once hull strength reached 0%, the ship is destroyed. 

Naturally, colliding with a planet or starship will have sufficient "yield" to immediately decrease the shields and hull to 0%.

Shields being raised affects how other systems work. For combat to be practical, the crew's weapons should penetrate their own shields. However, stealth fields, transporters, and tractor beams should not be usable with the shields raised. Communications and Probes are iffy, but traditionally they've worked while shields are raised.

### Crew and Flight Director Control

Shields are completely automated. The Flight Director should be able to see the current state of the hull and shields, and change them dynamically, but they don't need to watch too closely.

On the sensors screen, the crew and Flight Director should see the shield status as colored indicators around the ship icon. (The crew will have to scan specifically to see the shield status of other ships.)

Perhaps the Flight Director can be notified if the crew's death is imminent.

### Standing Questions
- Should functioning shields be a requirement for any kind of sub-light or warp travel, given that small particles would shred the ship at those speeds without space magic?
- Should shields actually be separated into 6 sections, or is that unnecessary complexity?

### Thorium Classic Shields Screenshot
![localhost_4442_client (47)](https://user-images.githubusercontent.com/6558157/163895355-0c256fb0-7ede-4db4-a59f-877ce757d930.jpg)

**Comment by The3gs** [comment]
My thoughts on the standing questions:
- I hadn't considered that shields might be required for such speeds. I think that this might be interesting, but I think it could also easily be explained away. Because warp drives are typically described as warping space, it might be that the warp drive can cooperate with sensors to specifically warp any obstacles out of the way. I certainly think that this would be worth considering.
- I think it would be more complex than is needed, and would be difficult for inexperienced players to know which shield are important for any given combat encounter, but I do think having something to manage related to the shields would be good. I think an alternative to having a shield for each side that will provide the shield's operators something to keep track of while still being fairly easy to know what to do in each situation would be to have 2 types of defense mechanisms. One mechanism, which we could call shields, would specialize against energy weapons, such as phasers. This would be the space-magic pulsing defense as described by Alex. The second, which I am thinking would be called deflectors, would instead specialize against physical attacks, such as torpedoes, deflecting them off course, possibly using warp technology. By having 2 mechanisms the operator would have to try to predict which types of weapons would be used against them in order to divert power in time. This would be much easier to reason about then what side the attacks are coming from, but still not as easy as "on or off."

**Reply by alexanderson1993** [reply]
The more I think about it, the less I like the idea of two passive defense systems. In practical terms, how does the crew interact with the "deflector" system any differently than the energy shield system? It seems like a slightly-different-but-still-the-same kind of thing, and I wonder how much value there to the crew is in redundant controls like this.

I really like the idea of having 6 independent shield segments because it integrates so well with the other members of the crew. Savvy Shields officers can tell the Pilot to rotate the ship so stronger shields receive the brunt of the impact. Maintenance teams will have to prioritize improving the efficiency of whichever shield is most necessary. And, of course, managing power to make sure the shields that are really needed are getting recharged as fast as possible.

In Thorium Classic, you can configure your shields to either be 6, 4, or solo configurations - I imagine we could do something similar with Thorium Nova. One big difference: Any operation on the shields affects all shields - changing the shield frequency and raising/lowering the shields. That should simplify the UI substantially.

**Reply by The3gs** [reply]
That's fair enough. It probably would come off as a bit redundant. I still feel like trying to communicate direction in 3d space could get messy, but I am sure you have much more experience in combat heavy campaigns than I do, so if you think it is practical I think that it would be fine.

The science side of my mind that likes the idea of carrying defences for different weapons intriguing, as it adds a bit more substance to the universe, though in practice I cannot think of a ton of ways to make it interesting for a player, so I'm not too offended if the idea gets dropped altogether.

**Reply by alexanderson1993** [reply]
I think it's worth considering more. There's certainly a lot of opportunity in terms of different kinds of ordinance having different effects and requiring different defenses. 

If you look back to https://github.com/Thorium-Sim/thorium-nova/discussions/224 you'll see there are lots of different kinds of defenses we can use - Countermeasures, signal jammer, railgun, [chaff launchers](https://github.com/Thorium-Sim/thorium-nova/discussions/234), and [ECM](https://github.com/Thorium-Sim/thorium-nova/discussions/233) are all possible ways we can add to defend against different kinds of attacks.

Also, there's nothing stopping us from adding a "deflector"-type shield in the future. For now, we'll stick with just one kind of shield.

Also, I imagine battles will be a little slower in Thorium Nova than other bridge sims. With proper alerts and indicators, it shouldn't be too hard for the crew to rotate their ship to take incoming hits on the sides with the strongest shields. But again, this is a tactic I only expect the most advanced crews to employ.

**Reply by Froginator45** [reply]
_Should functioning shields be a requirement for any kind of sub-light or warp travel, given that small particles would shred the ship at those speeds without space magic?_

I think that some sort of "shields" should be required for any sub-light travel. This could be the purpose of the "deflectors". Making it so that combat shields aren't needed for sub-light travel but the "deflector" is. It adds a little complexity to combat situations without making it too complicated.

_Should shields actually be separated into 6 sections, or is that unnecessary complexity?_

I think that having several sections of shields wouldn't be too hard to interact with if you make the integrity and damage control side of things the only parts that might be separate for each shield. I also don't see anyone being able to complain too loudly if you give the option to configure the number of shield emitters for each ship. Maybe the number of shields directly correlates with the number of emitters you have in the ship configuration. Therefore, smaller ships in general would have less and bigger would have more. If you were to do this, what would be the limit of emitters? 6 would account for most ship sizes. But stations like the Magellan might make sense to have more. The main thing this makes a little more complex would be the ship configuration: the FD would need to specify the emitter coverage area to get the physics to function correctly.

**Reply by Ryan-Anderson-PhD** [reply]
I think having different shield generators for different shield settings is very helpful for adding some tactical depth. I already do my best to simulate that using Thorium Classic. 

It makes sense to me that you need to have some kind of system to deal with space particles/dust. Of course, some of that depends on how you perceive FTL travel to occur. If we are going full on Alcubierre warp drive, then it would seem to me that the shields would be functioning inside the warp bubble, and would not be effective in anything that the warp field impacted along the way.

In my head, I have always sort of conceptualized either the subspace field or the jump drive envelope as being the "shields" around the warp bubble, and the thing that is responsible for handling any impacts between the warp field and anything in its path while FTL travel is occurring. In some ways, this is already simulated in Thorium Classic by being able to impact portions of the jump drive for jump drive sector stress. 

One reason I like the idea of having some other system being responsible for deflecting space dust and debris is because as things currently function, having the shields up drastically reduces the effectiveness of the stealth field. So, if shields are required to move at impulse or warp speed, then stealth at any speed besides thrusters is greatly limited. While I could see my you might go for that, it does limit story telling opportunities.

So, I might use something like a bussard ramscoop field for impulse speed and either a jump drive or subspace field as separate systems for dealing with dust and debris while traveling at sublight and FTL speeds.

Just my two cents.

**Comment by The3gs** [comment]
Alex and I have discussed several things regarding how shields should be represented as ship systems. We are thinking that there should be capability for multiple shield generators for the reasons discussed above, but we are less certain as to how we should define which shield generators would be effective by a given attack.

There are two main ways we can do this, with either there being a single shield generator allocated to each room, or having each room be influenced by multiple generators. 

With how we intend to implement the systems, each shield generator will be it's own system in it's own room. Which rooms are influenced by a generator would partially be a function of it's proximity to the room in question.

If we just have each room tied to a single generator, it would become very simple to answer whether a room is protected by a shield or not, as it would be a simple check of the single relevant generator. 
The rooms could be assigned a generator either by proximity, by manually assigning rooms, or by assigning a shape such as an ellipse or a cone to each generator where their effects lie.

Pros:
Easy calculations for effect of attacks.
Easy for users to know which locations are at risk of being hit

Cons:
Might require manual decisions about which generators effect which rooms.
Does not allow for areas which are affected by multiple shields, requiring hard cutoffs near boundaries.


If we were to allow multiple generators to protect a single room, we need to define a function which allocates to each shield a suitable portion of the damage. This adds considerable complexity, though it allows us to more smoothly model the shields, which has some added benefits such as targets near the middle of the ship being hit from various directions. 
Several functions could be used, but a simple one could be to have a shields benefit be inversely proportional to it's distance from the attack or the room.

Pros:
More general application with less manual adjustment.
No hard cutoffs between shield generators

Cons:
Might require more runtime checks and calculations
Requires finding a suitable function (Might not be that difficult, but I am not sure.)

Any other possible implementations are also more than welcome, so feel free to share your ideas.

**Reply by alexanderson1993** [reply]
I think one important thing to keep in mind is that not all ships will have "rooms" per se. Creating the deck maps is a time consuming process, and I think it should be possible to use ships in Thorium Nova that don't have deck maps or rooms. Also, single-room ships or very small ships, like fighters or shuttlecraft, also won't have rooms or deck maps.

That means any integration of rooms with the shields (and other systems) should be a progressive enhancement - it should be a nice improvement to the simulation for larger ships with deck maps, but it shouldn't be required for every ship.

**Reply by alexanderson1993** [reply]
If we take a step back and look at this from a first principles point of view, we need to ask "How does this make the game more interesting and enjoyable for crews and flight directors?" and "How does this make it possible to tell better stories?"

I think we've established that having multiple sections of shields is a great way to make the game more interesting for crews - the pilot can rotate the ship to direct incoming fire at the strongest shields, the power distribution officer can reroute power from one shield to another as needed - lots of great benefits.

When it comes to how damage is applied to rooms and systems, we might want to simplify it a bit. How many crews are going to care that when a weapon impacts the port side it only affects the rooms and systems on the port side of the ship? It might be simpler to calculate a coefficient of damage and _randomly_ apply it to different rooms and systems on the ship. It's not totally unbelievable, since the impact to the ship could cause reverberating damage to the rest of the ship, affecting the systems in different ways.

That would make the shields serve as just that - defensive measures that decrease the coefficient of damage as much as possible. The coefficient of damage is based on the strength of the shield that blocks the direction that the weapon came from.

**Reply by The3gs** [reply]
I've been thinking about it over the weekend and I think that we have a slightly different image in each of our mind as to what the shields will look like.

I picture that the shields are basically a bubble of protection, which protects anything within it's area. 

If I am understanding what you are saying correctly, it sounds like you are picturing the shields as protecting from attacks in a particular direction. Basically allocating each shield a portion of the sphere around the ship.

I suppose that my model was different from what was discussed in the original post, though I think there might still be some advantages to the bubble model. The biggest advantage I see would be that it would be easier to set up arbitrary shields for various ships, without having restrictions of how many shield generators are needed. For example, if it is based predominantly on direction, I cannot think of an easy way to allow for ship designers to use arbitrary shield arrangements, so we would probably have to restrict it to predefined patterns such as 1, 4 and 6. This would make the placement of multiple generators partially disconnected from the placement of the shields, which might not be a problem, but it depends what level of control we want to give ship designers. 
I picture the designing of a ships shields looking something like this:
![image](https://user-images.githubusercontent.com/7924239/166322767-1b7530a5-75b7-4536-b69a-44fbeae056bc.png)
with the dots being the shield generators, and the border representing the protected area. When a hit occurs, it would detect where the ship was hit, check the shield strength at that location, and then it would damage the shields and nearby systems.

Do you think we should discuss this option more, or would you rather stick with directional shield patterns?

**Reply by alexanderson1993** [reply]
That's a fascinating idea, and definitely more flexible than having a prescribed set of directional shield configurations (and yes, that was the intention - a 1/4/6 set of possible shield configurations)

Lets call this method "shield bubbles". Here are my questions:
- How does the system determine which shield was affected when a weapon impacts the ship?
- How do overlapping shields affect one another?
- Are the shield generators projecting a spherical shield, both above and below where the shield generator is? Or do they project a hemisphere, requiring shield generators on the top and bottom of the ship?

For that last question, I would prefer that they project a sphere. It simplifies the placement of the shields a lot.

Another simplification would be to keep the shield generator position itself separate from the Ship Map. Instead, you would place the shield generator on top of the 3D model of the ship and it would let you visualize the bubble that the various shield generators produce. You could change the radius of the bubble to make sure there is sufficient overlap between the shield generators.

Keeping the shield generator placement separate from the Ship Map would make this method work with any kind of ship, including small ships and ships that don't have a fully developed ship map.

I think my preference is still for directional shields, but the great thing about the way Thorium Nova's ship systems work is we could easily do _both_ - eventually, of course. Perhaps there are story points or strategies which make players prefer one option over the other, but they still have the choice. It also would create opportunities for variations in the different factions - wouldn't it be fun if you encountered an alien ship which used a totally different kind of shields than what your ship uses?

We still have to implement one type of shields first. Give some answers to my questions above and we'll go from there.

**Reply by The3gs** [reply]
> How does the system determine which shield was affected when a weapon impacts the ship?

I think it would be easiest to decide where the weapon would have hit first, and then determine if the point is protected by a given shield.

> How do overlapping shields affect one another?

In the 2d example image I sent above, all shields are combined into a single implicit function. The edges merge together similar to the concept of meta-balls (which is where I stole this equation from.) The border is where f(x,y)=0. To check if an area is protected by a shield I would first check if the solution to the equation for all active shields is greater than 1. If it is, the region is protected by the shield, and I would calculate each generator's effect and distribute damage according to their contribution to the field strength.

> Are the shield generators projecting a spherical shield, both above and below where the shield generator is? Or do they project a hemisphere, requiring shield generators on the top and bottom of the ship?

It would definitely be easier to have the shield be a full shell. I think it would be worth it to allow simple ellipsoid like shells as well, as that would allow the shield to match the ship a little closer without adding too much effort.

**Reply by alexanderson1993** [reply]
Ellipsoids add more complexity than you think, because now you have to track the elliptical scale in three dimensions plus the rotation in three dimensions. That’s not trivial.

My vote is still for directional shields. I think it would be easier to implement, it matches the previous versions of the controls, and it provides a well-known system for addressing shields - port, starboard, fore, etc etc. the bubble shields might be unexpectedly unfamiliar to crews, and both implementing them in code and designing them for ship builders seems more complicated. 

I would love to hear what other people think, though. I’ll include this discussion in the newsletter on Thursday and see if we can get some more thoughts.

**Reply by alexanderson1993** [reply]
@The3gs I think we've had enough discussion at this point. I haven't heard anything from anyone about this outside of this discussion. Go ahead and choose whichever shield style you think is best, and hit me up if you have any questions.

**Reply by The3gs** [reply]
I think for now I will stick with directional shields for now, because that seems to be what the most people would prefer, but I am not completely sure how we would allocate generators to each side of the shield. I don't think that it makes the most sense to have each side be required to have it's own generator as I think that would be overly restrictive on ship layouts, requiring ships with more complex shield configurations to have more generators.

What I am thinking is that you could have any generator be responsible for any number of shields, and if a specific generator goes offline, all associated shields would be shut down.

What do you think?

**Reply by alexanderson1993** [reply]
I can see it two ways:

1. Is your proposal. A ship is assigned N shield generator systems, and that system can be configured to cover any of the 6 cardinal directions. That means it possible two shield generators could overlap the same direction, which would give some redundancy. It scales down to a single generator covering the entire ship, 4 generators covering just the XZ sides of the ship, 6 generators with each covering every side, or more exotic configurations.

2. Do it like Thorium Classic, where you choose how many shields you want - 1, 4, or 6 - and it makes that many Shield systems for you.

I actually like the simplicity and flexibility of the first option. Let me know if that's not what you were describing, but if it is, go for that.

There's some more nuance to this - are you coming to the meeting tomorrow? If so, let's talk about it then. If not, I'll share some more thoughts tomorrow morning.

**Reply by The3gs** [reply]
I think that what you described is exactly what I am thinking.

I think that allowing arbitrary shield configurations would allow for many ships to be represented in whatever that make the most sense for their specific situation. 

So basically the shields officer would be given a control for each generator, the number of which would vary based on configuration?

**Reply by alexanderson1993** [reply]
That sounds right. It might be good to make a visualization of what the Shields officer's screen looks like. Would they raise and lower/set the frequency of each shield independently? Or would they all raise and lower/change frequency together? The latter is much simpler, and likely sufficient for most situations.

The complexity of the shields comes from routing power from one shield to another, which could get very interesting when you have different shields overlapping each other. 

The nuance I mentioned before is how we configure this on a ship. The way I've modeled it so far is that ship systems are defined separately from ships, and then you assign ship systems to the ships when you configure them. There currently isn't any way for you to set configuration options for ship systems on a per-ship basis, although that functionality could be added if we deemed it necessary. However, I would prefer to avoid it if possible to limit the complexity.

That gives us a few options for defining shields. 

1. Each shield segment is defined as a separate ship system. You would have to assign all of the necessary shield systems to a ship to make sure it has sufficient coverage. This seems like a bad plan.
2. When you create the ship system, you don't specify which direction(s) it covers. Instead, you configure the direction when you assign the system to the ship. This would have you assign multiple shield systems to a ship and then choose the direction for each of them. This is the situation I would prefer to avoid, but the tradeoffs aren't too bad.

There might be other options (let me know if you can think of any), but this does very quickly become complicated. For example, 
- should the power for a shield system be different if it covers several directions?
- It should be possible to rename each of the shield systems independently. How do we go about doing that?

As an aside... maybe having this much configuration is a bad idea 😅 

Right now I'm leaning towards option 2, which I'm not thrilled about, but seems necessary to implement this. Configuring shields would look like this:

- You first create a generic Shields ship system, which would include setting defaults for power and shield durability. You do _not_ configure direction at the phase. By default, the shields cover _all_ directions, which makes it easy to drop them into a ship without needing to configure it.
- You assign N number of shields to your ship. Here, you can override any of the ship system settings, and those overrides only apply to that ship. This is where you specify the direction the shields will go. This will also let you rename the ship system when it is assigned to the ship so you can properly name the shields "Fore Shields" or "Aft Defensive Grid" or whatever the user wants to name it.

It will take a bit to build out the architecture to configure ship systems on a per-ship basis, but as reluctant as I am to do it, I think it's for the best. Let me know if you agree, and I can start on the work soon.

**Comment by Ryan-Anderson-PhD** [comment]
I would vote for directional shields. While the other aspect is very
fascinating, it adds layers of complexity that might become a bit of a
mire.

**Comment by Froginator45** [comment]
This discussion has made me think of an app I have played: Space Arena. In that game you change the layout of your ship rooms to change how your combat would play out. If we were to apply something similar to Thorium Nova, the ship map could be something more defined and predetermined. Some rooms would have shields and therefore might be better on the outside of the ship. The ship could be reconfigured pre-mission and at certain locations with "mechanic shops". This would also give a different dimension to upgrades as each room could be upgraded individually (or collectively). An upgrade could add more shields, health, storage, etc. to each room. This might make scanning of enemy ships a bit more specific as a good scan might reveal the stats for each room of an enemy ship. With the outside rooms being more easily scanned and the inner rooms retaining a "fog-of-war" until better scans are completed or an outer room is sufficiently damaged. This might require more on the programming end but I think it would make the game play much more simplified while still retaining sufficient depth. I this thought completely changes some of the current thoughts but... Something to consider.

**Reply by alexanderson1993** [reply]
Interesting ideas. That might be worth a broader discussion on its own.

**Comment by jnoahjohnson** [comment]
I will just throw in my thoughts on the two questions

_Should functioning shields be a requirement for any kind of sub-light or warp travel, given that small particles would shred the ship at those speeds without space magic?_

I agree with the thoughts before that navigational "deflectors" should be a different system than shields. That being said, are there any implications with using shields to augment the deflectors? How would external objects, such as larger asteroids, be affected by ship with a shield system that is collides with it. Could you then theoretically move other objects?

_Should shields actually be separated into 6 sections, or is that unnecessary complexity?_

I like the points brought up about the tactical implications of having 6 sections. Could this also lead to "overloading" one side, leaving another side completely vulnerable or is the max power truly a max power? I wonder if there could be a "simple shield" mode as well depending on the age group or complexity of the mission.

One other consideration is the speed at which objects are moving. Look at how shields work in Dune. It is fascinating how the speed affects the use of shields. Not totally to the point, but something that can be explored as well.

**Comment by Ryan-Anderson-PhD** [comment]
This may be something where it would be wise to not over-engineer the
system. I might suggest one generator for fore, one for aft, one for
starboard, one for port, one for dorsal, and one for ventral. Or, you could
have a primary for each side with an auxiliary that can be used if the
primary is damaged, but which is less efficient. Something like that.

Sincerely,

Ryan Anderson, Ph.D., LMFT, MedFT
Co-founder, Telos U
Executive Director, Telos Discovery Space Center
He/Him
***@***.***
Cell: 801-505-3401
https://telos.org/
discoveryspacecenter.com


On Tue, May 10, 2022 at 4:18 PM Ben G. ***@***.***> wrote:

> I think for now I will stick with directional shields for now, because
> that seems to be what the most people would prefer, but I am not completely
> sure how we would allocate generators to each side of the shield. I don't
> think that it makes the most sense to have each side be required to have
> it's own generator as I think that would be overly restrictive on ship
> layouts, requiring ships with more complex shield configurations to have
> more generators.
>
> What I am thinking is that you could have any generator be responsible for
> any number of shields, and if a specific generator goes offline, all
> associated shields would be shut down.
>
> What do you think?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Thorium-Sim/thorium-nova/discussions/281#discussioncomment-2726007>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AWAOTB44ZBJUVPTT2D2EBC3VJLOBZANCNFSM5TXGXWPQ>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***
> com>
>

-- 

    *Telos is a* *Proud Member of Choose Mental Health 
<https://choosementalhealth.ck.page/7de19e5d70>*
